README.rst
=================
Flask API Builder
=================

.. image:: https://img.shields.io/pypi/v/flask_api_builder.svg?maxAge=2592000?style=plastic
   :alt: PyPI Version

.. image:: https://img.shields.io/pypi/l/flask_api_builder.svg?maxAge=2592000?style=plastic
   :alt: PyPI License

.. image:: https://travis-ci.org/Michael-F-Bryan/flask_api_builder.svg?branch=master
   :target: https://travis-ci.org/Michael-F-Bryan/flask_api_builder
   :alt: Current Travis Build

A really simple app that allows you to write out your RESTful APIs spec in a 
logical form and automatically generate stub functions for each endpoint.


Installing
==========

Install from Source
-------------------

To install from source, you will need to first clone the repository from
GitHub::

    git clone https://github.com/Michael-F-Bryan/flask_api_builder.git

Then run the `setup.py` file::

    python3 setup.py install


With Pip
--------

Or you can install directly from the Python Packaging Index::

    pip install flask_api_builder


Usage
=====

In order to generate your base API, you first need to write up a spec for it.
The spec contains the definitions for each of your API's end points, as well as
some variables to customize the Blueprint each API endpoint is attached to.


Writing The Spec File
---------------------

The spec file is designed to be quite intuitive to use. It is set out in a
table-like format, with each column being aligned with spaces. Any line
starting with a "#" is interpreted as a comment and ignored by the parser.


Defining Variables
~~~~~~~~~~~~~~~~~~

Defining variables is really easy, just add a line with `variable-name: value`.
The variable names currently supported are:

Blueprint-name
    The name for your blueprint (defaults to "api")
Prepend-with
    The path that all endpoints attached to your API Blueprint are prepended
    with (nothing by default)
Error-handlers
    Add pre-made 404 and 500 error handlers. They will be included unless
    something other than true, yes, 1, or y is entered (not case sensitive)

All variable names are **not case sensitive**.


Defining Endpoints
~~~~~~~~~~~~~~~~~~

API Endpoints are defined one per line, in a table-like format. Each line
consists of 3 or 4 fields delimited by two or more whitespace characters. This
means that you need to separate your columns by **at least** two spaces so the
parser will read them correctly.

The fields correspond to an endpoint's METHOD, URL, DESCRIPTION and
(optionally) the NAME of the function generated. If no name is provided then
one will be generated for you given the method and URL parameters (no
guarantees that the generated name will be intelligible though).


Here is an example of a simple spec for a to-do list app::

    blueprint-name: api
    #Method    URL                       Description
    GET       /tasks                     Retrieve list of tasks
    GET       /tasks/<int:task_id>       Retrieve task number <task_id>
    POST      /tasks                     Create a new task 



CHANGELOG.rst
Changelog
=========

[Unreleased]
------------

New Features
~~~~~~~~~~~~

-  Added error handlers to the generated code

Version 0.1.1 (2016-44-23)
--------------------------

New Features
~~~~~~~~~~~~

-  Added a commandline option to view an example spec
-  Turned the module into a package and gave it a commandline entry
   point

Documentation Changes
~~~~~~~~~~~~~~~~~~~~~

-  Wrote up a basic README
-  Added an example spec file

Administration and Chores
~~~~~~~~~~~~~~~~~~~~~~~~~

-  Added a makefile to automate the boring stuff
-  Turned the module into a package with a **main**.py file
-  Added the ability to create sdist's and wheels properly
 

Version 0.1.0 (2016-05-22)
--------------------------

New Features
~~~~~~~~~~~~

-  Copied the main file from a jupyter notebook and set up the project





tests/test_fab.py
import warnings
import pytest
import flask_api_builder as fab


@pytest.fixture
def f():
    url = '/tasks/foo'
    method = 'get'
    description = 'Do something'
    return fab.Function(url, method, description)


class TestFunction:
    def test_init_with_defaults(self):
        url = '/tasks/foo'
        method = 'get'
        description = 'Do something'
        f = fab.Function(url, method, description)
        assert f.url == url
        assert f.method == method.upper()
        assert description == f.description

    def test_init_with_custom_stuff(self):
        url = '/tasks/foo'
        method = 'get'
        description = 'Do something'
        blueprint = 'my_cool_api'
        name = 'some_restful_function'
        template = 'insert stuff here {}'
        f = fab.Function(url, method, description, blueprint, name, template)
        assert f.url == url
        assert f.method == method.upper()
        assert description == f.description
        assert f.blueprint == blueprint
        assert f.name == name
        assert f.template == template

    def test_get_args_with_no_args(self, f):
        got = f.get_args()
        should_be = []
        assert got == should_be

    def test_get_args_with_1_arg(self, f):
        f.url = '/todos/<int:todo_id>'
        got = f.get_args()
        should_be = ['todo_id']
        assert got == should_be

    def test_get_args_with_multiple_args(self, f):
        f.url = '/todos/<int:list_id>/items/<int:item_id>'
        got = f.get_args()
        should_be = ['list_id', 'item_id']
        assert got == should_be

    def test_get_args_with_no_type_declaration(self, f):
        f.url = '/todos/<todo_id>'
        got = f.get_args()
        should_be = ['todo_id']
        assert got == should_be

    def test_get_args_invalid_parameter(self, f):
        f.url = '/todos/<todo id>'
        with warnings.catch_warnings(record=True) as w:
            # Make sure all warnings are caught
            warnings.simplefilter("always")

            # Make sure zero parameters are found
            got = f.get_args()
            should_be = []
            assert got == should_be

            # Then make sure the warning was raised
            assert len(w) == 1
            assert issubclass(w[-1].category, fab.MalformedParameterWarning)
            assert 'The number of "<" or ">" is' in str(w[-1].message)


flask_api_builder.py
"""
A really simple app that allows you to write out your RESTful API's spec
in a logical form and automatically generate the necessary boilerplate.`
"""

from collections import namedtuple
import re
import warnings
import jinja2


class MalformedParameterWarning(Warning):
    """
    A warning printed when the parser thinks that a user wrote a malformed
    parameter.
    """


Rule = namedtuple('Rule', ('method', 'url', 'description', 'name'))

FUNCTION_TEMPLATE = jinja2.Template('''
@{{ blueprint }}.route('{{ url }}', methods=['{{ method|default('GET') }}'])
def {{ name }}({{ args_list|join(', ') }}):
    """
    {{ docstring }}
    """
    # TODO: Complete me!
    raise NotImplemented''')

PREAMBLE_TEMPLATE = jinja2.Template("""
from flask import Blueprint

{{ blueprint }} = Blueprint({{ bp_args|join(', ') }})
""")


class Function:
    """
    The class in charge of extracting all arguments from the url, generating
    a default name (if necessary), and rendering an individual REST endpoint
    function for Flask.
    """
    def __init__(self, url, method, description, blueprint='api',
                 name=None, template=None):
        self.url = url
        self.method = method.upper()
        self.description = description
        self.blueprint = blueprint
        self.name = name or None
        self.template = template or FUNCTION_TEMPLATE

    def get_args(self):
        """
        Find all the arguments in the url. These are usually bits
        that look like "<int:task_id>" and so on...
        """
        parameters = re.findall(r'<(?:[\w_]+:)?([\w_]+)>', self.url)

        if (self.url.count('<') != len(parameters) or
                self.url.count('<') != len(parameters)):
            warnings.warn(
                'The number of "<" or ">" is different from the number of '
                'parameters found. Make sure parameters look like '
                '"<int:task_id>"',
                MalformedParameterWarning)

        return parameters

    def generate_name(self, args):
        """
        Try to create a stock name for the function given the information
        provided.

        The idea is to create names like the following:
        * get_tasks
        * get_tasks_by_id
        * delete_task_by_id

        Most of the time you'll probably get gibberish, but there's a reason
        you are given the option of specifying a name in your spec.
        """
        name = []
        name.append(self.method.lower())

        # Get the last "word" in the url that isn't a parameter
        for word in reversed(self.url.split('/')):
            if '<' in word or '>' in word:
                continue
            else:
                name.append(word)
                break

        # If there are any arguments, then add "by_[last arg]"
        if args:
            name.append('by')
            name.append(args[-1])

        return '_'.join(name)

    def render(self):
        """
        Render the template given the information we already have.
        """
        args = self.get_args()
        name = self.name or self.generate_name(args)

        func = self.template.render(
            blueprint=self.blueprint,
            url=self.url,
            method=self.method,
            name=name,
            args_list=args,
            docstring=self.description)
        return func

    def __repr__(self):
        return '<{}: url="{}" method="{}">'.format(
            self.__class__.__name__,
            self.url,
            self.method)



class APIGenerator:
    """
    The generator which will render both the preamble and all the functions,
    then join it all together into one string.
    """
    def __init__(self, config, function_template=None):
        self.config = config
        self.blueprint = config.get('blueprint-name', 'api')
        self.rules = self.config['rules']
        self.function_template = function_template

    def preamble(self):
        """
        Generate the import statements and the blueprint definition.
        """
        bp_args = []
        bp_args.append(repr(self.blueprint))
        bp_args.append('__name__')

        if 'prepend-with' in self.config:
            bp_args.append('url_prefix="{}"'.format(self.config['prepend-with']))

        return PREAMBLE_TEMPLATE.render(blueprint=self.blueprint,
                                        bp_args=bp_args)

    def functions(self):
        """
        Turn each of the function "Rules" into their corresponding Function
        representations.
        """
        funcs = []
        for rule in self.rules:
            func = Function(
                rule.url,
                rule.method.upper(),
                rule.description,
                blueprint=self.blueprint,
                name=rule.name,
                template=self.function_template)
            funcs.append(func)
        return funcs

    def render(self):
        """
        Generate our API file.
        """
        lines = []

        # Add the preamble
        lines.append(self.preamble())

        # Give it a bit of space
        lines.append('')
        lines.append('')

        lines.extend(f.render()+'\n' for f in self.functions())

        return '\n'.join(lines).strip()


def parse_spec(spec):
    """
    Parse a spec line by line and create a config dictionary from the results.

    The parser will go through each line, splitting them wherever there are
    2 or more whitespace characters next to each other.

    Depending on the number of fields in the resulting list, the parser will
    populate the config dictionary accordingly.

    If there is only one field, assume that the user has provided some sort
    of assignment in the form "key: value", split on the ":" and then
    add the resulting key/value pair to the dictionary.

    If there are 3 fields, assume they are [METHOD, URL, DESCRIPTION] and
    add the corresponding rule to the dictionary.

    If there are 4 fields, assume they are [METHOD, URL, DESCRIPTION, NAME]
    and add the corresponding rule to the dictionary.

    Otherwise there is a syntax error.
    """
    config = {}
    config['rules'] = []

    for i, line in enumerate(spec.splitlines()):
        # Remove trailing whitespace
        line = line.strip()

        # Skip comments and empty lines
        if line.startswith('#') or not line:
            continue

        # Split by multiple space sections
        groups = re.split(r'\s\s+', line)

        if len(groups) == 1:
            # It's an option
            left, right = groups[0].split(':')

            # If it fails, throw syntax error
            config[left.strip().lower()] = right.strip()
        elif len(groups) == 3:
            method, url, description = groups
            new_rule = Rule(method, url, description, None)
            config['rules'].append(new_rule)
        elif len(groups) == 4:
            method, url, description, name = groups
            new_rule = Rule(method, url, description, name)
            config['rules'].append(new_rule)
        else:
            raise SyntaxError('Too many fields on line {}'.format(i+1))

    return config


def make_api(spec):
    """
    Given a spec, parse the spec, make the API generator, and then render
    an API for the user.
    """
    cfg = parse_spec(spec)
    generator = APIGenerator(cfg)
    return generator.render().strip()


flask_api_builder/__main__.py
"""
Generate the basic functionality required to create a Flask RESTful API given
a spec file written in the appropriate format.

Usage: flask_api_builder [options] <spec_file>

Options:
    -o=OUT --out-file=OUT           The path of the generated spec
                                    [Default: stdout]
    -e --example                    Shows an example spec file
    -h --help                       Print this help text
    -V --version                    Print the version information
"""

import os
import docopt
import flask_api_builder as fab


def _main(args):
    spec_file = os.path.abspath(args['<spec_file>'])
    spec = open(spec_file).read()

    api_as_str = fab.make_api(spec)

    if args['--out-file'] == 'stdout':
        print(api_as_str)
    else:
        with open(args['--out-file'], 'w') as f:
            f.write(api_as_str)


def main():
    # We need to do this mucking around with main() and _main() so that
    # you can use the API generator with both the "python -m" syntax and from
    # the commandline entry point
    args = docopt.docopt(__doc__, version=fab.__version__)
    _main(args)


if __name__ == "__main__":
    main()


flask_api_builder/__init__.py
"""
A really simple app that allows you to write out your RESTful API's spec
in a logical form and automatically generate the necessary boilerplate.`
"""

__version__ = '0.1.1'
__all__ = [
        'MalformedParameterWarning',
        'Function',
        'APIGenerator',
        'make_api',
        'parse_spec',
        'Rule',
]


from collections import namedtuple
import re
import warnings
import jinja2


class MalformedParameterWarning(Warning):
    """
    A warning printed when the parser thinks that a user wrote a malformed
    parameter.
    """


Rule = namedtuple('Rule', ('method', 'url', 'description', 'name'))

FUNCTION_TEMPLATE = jinja2.Template('''
@{{ blueprint }}.route('{{ url }}', methods=['{{ method|default('GET') }}'])
def {{ name }}({{ args_list|join(', ') }}):
    """
    {{ docstring }}
    """
    # TODO: Complete me!
    raise NotImplemented''')

PREAMBLE_TEMPLATE = jinja2.Template("""
from flask import Blueprint, jsonify

{{ blueprint }} = Blueprint({{ bp_args|join(', ') }})
""")

ERROR_HANDLER_TEMPLATE = jinja2.Template("""
@{{ blueprint }}.errorhandler(404)
def page_not_found(error):
    response = jsonify({'error': 'not found'})
    response.status_code = 404
    return response

@{{ blueprint }}.errorhandler(500)
def server_error(error):
    response = jsonify({'error': 'server error'})
    response.status_code = 500
    return response
""")


def _is_true(string):
    """
    Check if something is truthy.
    """
    return str(string.lower().strip()) in ['true', '1', 'yes', 'y']


class Function:
    """
    The class in charge of extracting all arguments from the url, generating
    a default name (if necessary), and rendering an individual REST endpoint
    function for Flask.
    """
    def __init__(self, url, method, description, blueprint='api',
                 name=None, template=None):
        self.url = url
        self.method = method.upper()
        self.description = description
        self.blueprint = blueprint
        self.name = name or None
        self.template = template or FUNCTION_TEMPLATE

    def get_args(self):
        """
        Find all the arguments in the url. These are usually bits
        that look like "<int:task_id>" and so on...
        """
        parameters = re.findall(r'<(?:[\w_]+:)?([\w_]+)>', self.url)

        if (self.url.count('<') != len(parameters) or
                self.url.count('<') != len(parameters)):
            warnings.warn(
                'The number of "<" or ">" is different from the number of '
                'parameters found. Make sure parameters look like '
                '"<int:task_id>"',
                MalformedParameterWarning)

        return parameters

    def generate_name(self, args):
        """
        Try to create a stock name for the function given the information
        provided.

        The idea is to create names like the following:
        * get_tasks
        * get_tasks_by_id
        * delete_task_by_id

        Most of the time you'll probably get gibberish, but there's a reason
        you are given the option of specifying a name in your spec.
        """
        name = []
        name.append(self.method.lower())

        # Get the last "word" in the url that isn't a parameter
        for word in reversed(self.url.split('/')):
            if '<' in word or '>' in word:
                continue
            else:
                name.append(word)
                break

        # If there are any arguments, then add "by_[last arg]"
        if args:
            name.append('by')
            name.append(args[-1])

        return '_'.join(name)

    def render(self):
        """
        Render the template given the information we already have.
        """
        args = self.get_args()
        name = self.name or self.generate_name(args)

        func = self.template.render(
            blueprint=self.blueprint,
            url=self.url,
            method=self.method,
            name=name,
            args_list=args,
            docstring=self.description)
        return func

    def __repr__(self):
        return '<{}: url="{}" method="{}">'.format(
            self.__class__.__name__,
            self.url,
            self.method)



class APIGenerator:
    """
    The generator which will render both the preamble and all the functions,
    then join it all together into one string.
    """
    def __init__(self, config, function_template=None):
        self.config = config
        self.blueprint = config.get('blueprint-name', 'api')
        self.rules = self.config['rules']
        self.function_template = function_template

    def preamble(self):
        """
        Generate the import statements and the blueprint definition.
        """
        bp_args = []
        bp_args.append(repr(self.blueprint))
        bp_args.append('__name__')

        if 'prepend-with' in self.config:
            bp_args.append('url_prefix="{}"'.format(self.config['prepend-with']))

        return PREAMBLE_TEMPLATE.render(blueprint=self.blueprint,
                                        bp_args=bp_args)

    def functions(self):
        """
        Turn each of the function "Rules" into their corresponding Function
        representations.
        """
        funcs = []
        for rule in self.rules:
            func = Function(
                rule.url,
                rule.method.upper(),
                rule.description,
                blueprint=self.blueprint,
                name=rule.name,
                template=self.function_template)
            funcs.append(func)
        return funcs

    def error_handlers(self):
        """
        Make a couple useful error handler routes for the user.
        """
        return ERROR_HANDLER_TEMPLATE.render(blueprint=self.blueprint)


    def render(self):
        """
        Generate our API file.
        """
        lines = []

        # Add the preamble
        lines.append(self.preamble())

        # Give it a bit of space
        lines.append('')
        lines.append('')

        if _is_true(self.config.get('error-handlers', 'true')):
            # Add the error handlers
            lines.append(self.error_handlers())

            # Give it a bit of space
            lines.append('')
            lines.append('')

        lines.extend(f.render()+'\n' for f in self.functions())

        return '\n'.join(lines).strip()


def parse_spec(spec):
    """
    Parse a spec line by line and create a config dictionary from the results.

    The parser will go through each line, splitting them wherever there are
    2 or more whitespace characters next to each other.

    Depending on the number of fields in the resulting list, the parser will
    populate the config dictionary accordingly.

    If there is only one field, assume that the user has provided some sort
    of assignment in the form "key: value", split on the ":" and then
    add the resulting key/value pair to the dictionary.

    If there are 3 fields, assume they are [METHOD, URL, DESCRIPTION] and
    add the corresponding rule to the dictionary.

    If there are 4 fields, assume they are [METHOD, URL, DESCRIPTION, NAME]
    and add the corresponding rule to the dictionary.

    Otherwise there is a syntax error.
    """
    config = {}
    config['rules'] = []

    for i, line in enumerate(spec.splitlines()):
        # Remove trailing whitespace
        line = line.strip()

        # Skip comments and empty lines
        if line.startswith('#') or not line:
            continue

        # Split by multiple space sections
        groups = re.split(r'\s\s+', line)

        if len(groups) == 1:
            # It's an option
            left, right = groups[0].split(':')

            # If it fails, throw syntax error
            config[left.strip().lower()] = right.strip()
        elif len(groups) == 3:
            method, url, description = groups
            new_rule = Rule(method, url, description, None)
            config['rules'].append(new_rule)
        elif len(groups) == 4:
            method, url, description, name = groups
            new_rule = Rule(method, url, description, name)
            config['rules'].append(new_rule)
        else:
            raise SyntaxError('Too many fields on line {}'.format(i+1))

    return config


def make_api(spec):
    """
    Given a spec, parse the spec, make the API generator, and then render
    an API for the user.
    """
    cfg = parse_spec(spec)
    generator = APIGenerator(cfg)
    return generator.render().strip()
